File: ainpc-core-plugin/src/test/java/ro/ainpc/engine/FeaturePackMetadataValidatorTest.java
package ro.ainpc.engine;

import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.YamlConfiguration;
import org.junit.jupiter.api.Test;
import ro.ainpc.platform.RuntimeMode;

import java.io.File;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class FeaturePackMetadataValidatorTest {

    @Test
    void acceptsDeclaredAddonMetadataForCurrentRuntime() throws Exception {
        FeaturePackMetadataValidator.ValidationResult result = FeaturePackMetadataValidator.validate(load("""
            id: demo_pack
            name: "Demo Pack"
            addon:
              type: "scenario"
              version: "1.0.0"
              primary_scenario: true
              runtime_modes: ["standalone", "hybrid"]
              capabilities: ["scenarios", "progression"]
              dependencies: ["core_demo"]
            """), new File("demo_pack.yml"), RuntimeMode.STANDALONE);

        assertTrue(result.valid());
        assertTrue(result.errors().isEmpty());
    }

    @Test
    void rejectsUnknownRuntimeMode() throws Exception {
        FeaturePackMetadataValidator.ValidationResult result = FeaturePackMetadataValidator.validate(load("""
            id: demo_pack
            addon:
              type: "feature"
              runtime_modes: ["future_mode"]
            """), new File("demo_pack.yml"), RuntimeMode.STANDALONE);

        assertFalse(result.valid());
        assertTrue(result.errors().stream().anyMatch(error -> error.contains("runtime necunoscut")));
    }

    @Test
    void rejectsPackThatDoesNotSupportCurrentRuntime() throws Exception {
        FeaturePackMetadataValidator.ValidationResult result = FeaturePackMetadataValidator.validate(load("""
            id: advanced_pack
            addon:
              type: "feature"
              runtime_modes: ["advanced"]
            """), new File("advanced_pack.yml"), RuntimeMode.STANDALONE);

        assertFalse(result.valid());
        assertTrue(result.errors().stream().anyMatch(error -> error.contains("runtime-ul curent")));
    }

    @Test
    void rejectsInvalidAddonListMetadata() throws Exception {
        FeaturePackMetadataValidator.ValidationResult result = FeaturePackMetadataValidator.validate(load("""
            id: demo_pack
            addon:
              type: "feature"
              capabilities: "scenarios"
            """), new File("demo_pack.yml"), RuntimeMode.STANDALONE);

        assertFalse(result.valid());
        assertTrue(result.errors().stream().anyMatch(error -> error.contains("capabilities trebuie sa fie lista")));
    }

    private YamlConfiguration load(String content) throws InvalidConfigurationException {
        YamlConfiguration configuration = new YamlConfiguration();
        configuration.loadFromString(content);
        return configuration;
    }
}

File: ainpc-core-plugin/src/test/kotlin/ro/ainpc/engine/FeaturePackMetadataValidatorTest.kt
package ro.ainpc.engine

import org.bukkit.configuration.InvalidConfigurationException
import org.bukkit.configuration.file.YamlConfiguration
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import ro.ainpc.platform.RuntimeMode
import java.io.File

class FeaturePackMetadataValidatorTest {
    @Test
    @Throws(Exception::class)
    fun acceptsDeclaredAddonMetadataForCurrentRuntime() {
        val result = FeaturePackMetadataValidator.validate(
            load(
                """
                id: demo_pack
                name: "Demo Pack"
                addon:
                  type: "scenario"
                  version: "1.0.0"
                  primary_scenario: true
                  runtime_modes: ["standalone", "hybrid"]
                  capabilities: ["scenarios", "progression"]
                  dependencies: ["core_demo"]
                """
            ),
            File("demo_pack.yml"),
            RuntimeMode.STANDALONE
        )

        assertTrue(result.valid())
        assertTrue(result.errors().isEmpty())
    }

    @Test
    @Throws(Exception::class)
    fun rejectsUnknownRuntimeMode() {
        val result = FeaturePackMetadataValidator.validate(
            load(
                """
                id: demo_pack
                addon:
                  type: "feature"
                  runtime_modes: ["future_mode"]
                """
            ),
            File("demo_pack.yml"),
            RuntimeMode.STANDALONE
        )

        assertFalse(result.valid())
        assertTrue(result.errors().any { error -> error.contains("runtime necunoscut") })
    }

    @Test
    @Throws(Exception::class)
    fun rejectsPackThatDoesNotSupportCurrentRuntime() {
        val result = FeaturePackMetadataValidator.validate(
            load(
                """
                id: advanced_pack
                addon:
                  type: "feature"
                  runtime_modes: ["advanced"]
                """
            ),
            File("advanced_pack.yml"),
            RuntimeMode.STANDALONE
        )

        assertFalse(result.valid())
        assertTrue(result.errors().any { error -> error.contains("runtime-ul curent") })
    }

    @Test
    @Throws(Exception::class)
    fun rejectsInvalidAddonListMetadata() {
        val result = FeaturePackMetadataValidator.validate(
            load(
                """
                id: demo_pack
                addon:
                  type: "feature"
                  capabilities: "scenarios"
                """
            ),
            File("demo_pack.yml"),
            RuntimeMode.STANDALONE
        )

        assertFalse(result.valid())
        assertTrue(result.errors().any { error -> error.contains("capabilities trebuie sa fie lista") })
    }

    @Throws(InvalidConfigurationException::class)
    private fun load(content: String): YamlConfiguration {
        val configuration = YamlConfiguration()
        configuration.loadFromString(content)
        return configuration
    }
}
